In [6]:
import numpy as np
from exputils.st_norm import compute_st_norm
from exputils.RoM.fwht import calculate_RoM_FWHT
from exputils.state.random import make_random_quantum_state
from tqdm.auto import tqdm

n_qubit = 10
SEED_CNT = 100

RoMs = []
st_norms = []
ratios = []
ratios_per_qubit = []

for seed in tqdm(range(SEED_CNT)):
    rho_vec = make_random_quantum_state("mixed", n_qubit, seed)
    st_norm = compute_st_norm(rho_vec)
    assert np.isclose(st_norm, np.linalg.norm(rho_vec, ord=1) / (2**n_qubit))
    RoM = calculate_RoM_FWHT(n_qubit, rho_vec)[0]
    RoMs.append(RoM)
    st_norms.append(st_norm)
    ratios.append(RoM / st_norm)
    ratios_per_qubit.append((RoM / st_norm) ** (1 / n_qubit))

print(f"{np.mean(RoMs)} \pm {np.std(RoMs)}")
print(f"{np.mean(st_norms)} \pm {np.std(st_norms)}")
print(f"{np.mean(ratios)} \pm {np.std(ratios)}")
print(f"{np.mean(ratios_per_qubit)} \pm {np.std(ratios_per_qubit)}")

  0%|          | 0/100 [00:00<?, ?it/s]

25.545793517961602 \pm 0.018299047956833927
0.7988283438456045 \pm 0.0005544557687705419
31.97907948964125 \pm 0.012625191412967179
1.414121068777219 \pm 5.58242209086292e-05


for ten qubits random mixed quantum state,

$$
    \mathcal{R}(\rho) =25.546 (\text{SD: } 1.8300 \times 10^{-2}) \\
    \left(\frac{\mathcal{R}(\rho)}{||\rho||_{\text{st}}}\right) ^ {1/n}=
    1.41412 (\text{SD: } 5.5824 \times 10^{-5})
$$
